Improved parsing of PyNEST test results and list of failed tests

heplesser opened the pull request:
This fixes #1548 and provides a list of failed tests as part of the Testsuite summary.

This PR changes that way in which results from Python tests run with nosetests are parsed. This solves the problem described in #1548. Since that problem makes it impossible to run the complete testssuite on affected systems, I prioritise this as "high". The PR also provides a list of failed tests.

Some points to consider:
- Python test results are now parsed from the `pynest_tests.xml` file using [junitparser](https://pypi.org/project/junitparser/), a pip-installable dependency.
- Failing tests are not included in the list of failed tests if
    - Python tests were run without nosetests
    - they occurred during the still-experimental C++ test phase.

Here is an example of the resulting list (illustrative example):

```
NEST Testsuite Summary
----------------------
  NEST Executable: /Users/plesser/NEST/code/bld_gcc_mpi/install/bin/nest
  SLI Executable : /Users/plesser/NEST/code/bld_gcc_mpi/install/bin/sli
  Total number of tests: 920
     Passed: 903
     Skipped: 14 (8 PyNEST)
     Failed: 3 (1 PyNEST)

***
*** The following tests failed while running the NEST Testsuite:
***
***     unittests/test_gif_cond_exp.sli
***     regressiontests/issue-779-1016.py
***     nest.tests.test_connect_distributions.TestDists.testBinomialClippedDist
***
*** Please report the problem at ...
```

jougs commented on testsuite/do_tests.sh.in:
```diff
@@ -732,11 +735,17 @@ if test "x${TEST_PYNEST}" = xtrue ; then
-        PYNEST_TEST_SKIPPED="$(  tail -n 1 ${TEST_LOGFILE} | sed -$EXTENDED_REGEX_PARAM 's/.*SKIP=([0-9]+).*/\1/')"
-        PYNEST_TEST_FAILURES="$( tail -n 3 ${TEST_LOGFILE} | grep \( | sed -$EXTENDED_REGEX_PARAM 's/^[a-zA-Z]+ \((.*)\)/\1/' | sed -$EXTENDED_REGEX_PARAM 's/.*failures=([0-9]+).*/\1/' )"
-        PYNEST_TEST_ERRORS="$( tail -n 3 ${TEST_LOGFILE} | grep \( | sed -$EXTENDED_REGEX_PARAM 's/^[a-zA-Z]+ \((.*)\)/\1/' | sed -$EXTENDED_REGEX_PARAM 's/.*errors=([0-9]+).*/\1/' )"
+        ntsummary=`@PYTHON@ -c "import junitparser as jp; r = jp.JUnitXml.fromfile('${TEST_OUTDIR}/pynest_tests.xml'); pl = ['.'.join((s.classname, s.name)) for s in r if s.result and not isinstance(s.result, jp.junitparser.Skipped)]; print(r.tests, r.skipped, r.failures, r.errors, pl)"`
```
I would prefer if the executed code were put in its own script and given the `TEST_OUTDIR` as an argument. Also, I'd find it nicer if the execution of the subshell were done using the `$( ... )` syntax intsead of <code>\`...\`</code>, as this is also used (almost) everywhere else in the script.

jougs commented on testsuite/do_tests.sh.in:
```diff
@@ -847,7 +856,13 @@ echo
-    echo "*** There were errors detected during the run of the NEST test suite!"
+    echo "*** The following tests failed while running the NEST Testsuite:"
+    echo "***"
+    for ft in ${FAILED_TESTS}
+    do
```
```suggestion
    for ft in ${FAILED_TESTS}; do
```

jougs commented on testsuite/do_tests.sh.in:
```diff
@@ -847,7 +856,13 @@ echo
+    do
+      echo "***     ${ft}"
+    done
+    echo "***"
```
This line is being printed twice if there were no failing tests. However, I'm not sure if this justifies an additional `if`... If you go for it, I'd like to see its body to also include the `for` loop.

heplesser commented:
@jougs I have now revised parsing completely, leaving reporting entirely to a python script. That gives a neat table like this:
```
NEST Testsuite Results

----------------------------------------------------------------------
Phase                    Tests   Skipped  Failures    Errors      Time
----------------------------------------------------------------------
01 basic tests               6         0         0         0       8.0
02 selftests                 8         0         0         0      20.0
03 unittests               186         0         0         0     113.0
04 regressiontests          91         0         1         0      73.0
05 mpitests                 77         0         0         0     119.0
07 pynest tests            747         8         0         0     104.6
08 cpptests                 19         0         0         2       0.0
----------------------------------------------------------------------
Total                     1134         8         1         2     437.6
----------------------------------------------------------------------

THE NEST TESTSUITE DISCOVERED PROBLEMS
    The following tests failed:
        regressiontests.issue-779-1016.py.regressiontests/issue-779-1016.py
        test_blockvector.test_size
        test_blockvector.test_random_access

    Please report test failures by creating an issue at
        https://github.com/nest/nest_simulator/issues
```

Integrating that with your file reorganisation in #1580 should not be difficult. I also fixed a few other things, including time-taking on macOS, and removed some output that provided no useful information but gave a few details away about the machine on which it was run. One should probably add some information about NEST source code version etc to the report.

jougs commented on testsuite/do_tests.sh.in:
```diff
@@ -541,64 +466,20 @@ echo
-else
-    rm -rf "${TEST_TMPDIR}"
-fi
+@PYTHON@ "$(dirname $0)/summarize_tests.py" "${TEST_OUTDIR}"
```
Probably you already know this, but anyway: I think the current failure in Travis is caused by `@PYTHON@` being empty in case we compiled without PyNEST, which leads to `summarize_tests.py` (incorrectly) being executed as a shell script due to a missing `#!/usr/bin/env python3` shebang in that file.

heplesser replied:
Indeed. Interestingly, `parse_travis_log.py` also has no shebang and is invoked as `python extras/parse_travis_log.py travis_build.sh.log ;` from `travis.yml`.  What would be the most consistent and portable approach to running Python scripts during the build/test process?

jougs replied:
I think invoking as `python ...` is perfectly fine for running helper scripts as we don't require any specific Python version for them to do their job. `@PYTHON@` should only be used to run the actual Python tests, as this ensures that we actually use the exact version of Python that was used to compile against. Insofar, leaving out the shebang would be perfectly fine from my point of view.

heplesser commented on .travis.yml:
```diff
@@ -170,7 +170,7 @@ before_install:
          # terminaltables is required by parse_travis_log.py to create
          # the build summary.
-         pip install cython scipy matplotlib terminaltables
+         pip install cython scipy matplotlib terminaltables nose junitparser
```
... but `apt` doesn't work on macOS. But I should then remove `nose` from the `apt` above.

jougs replied:
Perfect!

heplesser commented:
@jougs Thanks for your comments and one further question: I changed the shebang in `do_tests.sh.in` from `!#/bin/sh` to `!#/bin/bash` since I did not find a way to make the [substitutions in `time_cmd()`](https://github.com/nest/nest-simulator/blob/c4891aaa7921b43ce9e88e62997f410141fa1600/testsuite/junit_xml.sh#L48) work with the `sh` on Travis (I haven't even been able to find out what `sh` precisely that is and how to run it on my computer for easier debugging). Do you see this as a problem?

jougs commented:
No, I don't see a problem with moving to `bash`.

That said, all these slighly incompatible shells and the complexity involved in using them properly for writing JunitXML, doing calculations, running subprocesses, parsing results and getting proper reporting leaves me wondering if we should not "just" replace the whole test harness by one written in Python.

After all, the reason for not using Python in the first place was that it was not widely available back in the day when we wrote the first version of the testsuite. As this is not true anymore, we could as well just make our lifes easy.

What do you think about this? My idea would be to continue the modularization efforts I already have in the pipeline and then successively replace parts by Python code.

heplesser commented:
> No, I don't see a problem with moving to `bash`.

Good.

> That said, all these slighly incompatible shells and the complexity involved in using them properly for writing JunitXML, doing calculations, running subprocesses, parsing results and getting proper reporting leaves me wondering if we should not "just" replace the whole test harness by one written in Python.

Indeed!

> What do you think about this? My idea would be to continue the modularization efforts I already have in the pipeline and then successively replace parts by Python code.

I think that switching to Python would make a lot of sense. Would you create a separate discussion issue about it and put it on the agenda of the next Open VC? We should then stick to a conservative Python, compatible with Py 3.5 and as few non-standard-lib packages as possible.

lekshmideepu commented on .travis.yml:
```diff
@@ -158,20 +159,28 @@ before_install:
-           brew cask list
+           # brew list
+           # brew leaves
+           # brew cask list
```
Probably you could remove the commented lines.

heplesser replied:
Done!

lekshmideepu replied:
Perfect! Thanks

jougs commented on .travis.yml:
```diff
@@ -158,20 +159,25 @@ before_install:
+           pip install nose-1.3.7noman-py3-none-any.whl
          fi
-
+        
```
```suggestion

```

jougs commented on testsuite/do_tests.sh.in:
```diff
@@ -462,73 +441,19 @@ if test "x${TEST_PYNEST}" = xtrue ; then
     #    dummy directory to search for tests.
 
-    if command -v @NOSETESTS@ >/dev/null 2>&1 && @PYTHON@ @NOSETESTS@ --with-xunit --xunit-file=/dev/null --where="${TEST_OUTDIR}" >/dev/null 2>&1; then
+     if command -v @NOSETESTS@ >/dev/null 2>&1 && @PYTHON@ @NOSETESTS@ --with-xunit --xunit-file=/dev/null --where="${TEST_OUTDIR}" >/dev/null 2>&1; then
```
```suggestion
    if command -v @NOSETESTS@ >/dev/null 2>&1 && @PYTHON@ @NOSETESTS@ --with-xunit --xunit-file=/dev/null --where="${TEST_OUTDIR}" >/dev/null 2>&1; then
```

jougs commented on testsuite/do_tests.sh.in:
```diff
@@ -462,73 +441,19 @@ if test "x${TEST_PYNEST}" = xtrue ; then
         echo
-        echo "  Nosetests unavailable. Using fallback test harness."
-        echo "  Fewer tests will be excuted."
+        echo "  Not running pynest tests because nosetests is missing."
```
```suggestion
        echo "  Not running PyNEST tests because nosetests is not available."
```

jougs commented:
@heplesser: Now that #1578 is finally running fine, could you please have a quick look at my suggestions, so we can merge this before #1578?

heplesser commented:
@jougs I have accepted all your suggestions. Once the tests have passed, this is thus ready to be merged.

jougs commented:
Awesome, thanks!
